legend-engine-core/legend-engine-core-executionPlan-execution/legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/nodes/helpers/freemarker/FreeMarkerExecutor.java

```diff
@@ -79,17 +86,68 @@ else if (v instanceof StreamingResult)
         return StringUtils.isBlank(templateFunctions) ? process(input, variableMap, templateFunctions) : processRecursively(input, variableMap, templateFunctions);
     }
 
+
+    private static boolean isPlaceHolder(Object object)
+    {
+        return object instanceof String && pattern.matcher((String)object).find();
+    }
+
+    private static class TemplateHashModelOverride implements TemplateHashModel
+    {
+        private Map map;
+        private String templateFunctions;
+
+        TemplateHashModelOverride(Map map, String templateFunctions)
+        {
+            this.map = map;
+            this.templateFunctions = templateFunctions;
+        }
+
+        @Override
+        public TemplateModel get(String s) throws TemplateModelException
+        {
+            Object result = map.get(s);
+            boolean isResultUnchanged = false;
+            while (!isResultUnchanged && isPlaceHolder(result))
+            {
+                String lastProcessedResult = (String) result;
```

Comment:
are we sure this is safe? if so we should change the map to Map<String, String>, in fact the map passed in is: Map<String, ?> .... probably just needs to be tidied up but let's do that.

Reply:
We require <String, ?> because there are instances where the map contains string key with sone class Instance as a value to , and they go through the processRecursively call to create template functions.  It is only when we are processing placeholders, we find a Map of key strings and value of type string